In [ ]:
# ======================================================
# Noise Schedule Utilities
# ======================================================

import torch
import matplotlib.pyplot as plt

# -------------------------
# 1) Define Noise Schedule
# -------------------------
def make_beta_schedule(T, beta_start=1e-4, beta_end=0.02):
    """
    Linear beta schedule exactly as in original DDPM code
    """
    return torch.linspace(beta_start, beta_end, T)

def build_noise_schedule(T, device="cpu"):
    """
    Returns betas, alphas, alpha_bars
    Compatible with main diffusion training code
    """
    betas = make_beta_schedule(T).to(device)
    alphas = (1.0 - betas).to(device)
    alpha_bars = torch.cumprod(alphas, dim=0).to(device)
    return betas, alphas, alpha_bars


# -------------------------
# 2) Inspect Schedule
# -------------------------
def print_schedule_samples(betas, alphas, alpha_bars, steps=[0,1,2,3,4,5,6,7,8,9,10, 50, 100, 200, 300,400,500]):
    print(" t |    beta    |   alpha   | alpha_bar")
    print("----------------------------------------")
    for t in steps:
        print(f"{t:3d} | {betas[t]:.6f} | {alphas[t]:.6f} | {alpha_bars[t]:.6f}")


# -------------------------
# 3) Quick Test Run
# -------------------------
if __name__ == "__main__":
    T = 500
    device = "cuda" if torch.cuda.is_available() else "cpu"

    betas, alphas, alpha_bars = build_noise_schedule(T, device)

    print_schedule_samples(betas, alphas, alpha_bars)


 t |    beta    |   alpha   | alpha_bar
----------------------------------------
  0 | 0.000100 | 0.999900 | 0.999900
  1 | 0.000140 | 0.999860 | 0.999760
  2 | 0.000180 | 0.999820 | 0.999580
  3 | 0.000220 | 0.999780 | 0.999361
  4 | 0.000260 | 0.999740 | 0.999101
  5 | 0.000299 | 0.999701 | 0.998802
  6 | 0.000339 | 0.999661 | 0.998464
  7 | 0.000379 | 0.999621 | 0.998085
  8 | 0.000419 | 0.999581 | 0.997667
  9 | 0.000459 | 0.999541 | 0.997209
 10 | 0.000499 | 0.999501 | 0.996711
 50 | 0.002094 | 0.997906 | 0.945552
100 | 0.004088 | 0.995912 | 0.809140
200 | 0.008076 | 0.991924 | 0.438713
300 | 0.012064 | 0.987936 | 0.159121
400 | 0.016052 | 0.983948 | 0.038544


IndexError: index 500 is out of bounds for dimension 0 with size 500

In [ ]:
1(image without noise)- 0.000100(beta) = 0.999100 (alpha)

beta كمية ال noise الي ضفناها على الصورة
alpha قديش باقي من الصورة بعد اضافة ال noise  ل t 1

alpha_bar
تمثل قديش باقي من الصورة بعد عدد معين من t

SyntaxError: invalid syntax. Perhaps you forgot a comma? (ipython-input-1480893059.py, line 1)

In [ ]:
تم


---

## What does each term represent?

- **beta (βₜ)**  
  The amount of **new noise added** at timestep `t`.

- **alpha (αₜ = 1 − βₜ)**  
  The amount of **signal preserved** at timestep `t`.

- **alpha_bar (ᾱₜ)**  
  The **total remaining fraction of the original image** after `t` steps.

---

## Interpreting the Results

### 🔹 At t = 0
- `alpha_bar ≈ 0.9999`
- Almost the entire original image is still visible.
- Very tiny amount of noise added.

---

###  At t = 50
- `alpha_bar ≈ 0.9455`
- About **94% of the original image remains**.
- Noise is present but image structure is still clear.

---

###  At t = 100
- `alpha_bar ≈ 0.8091`
- About **81% of the image remains**.
- Noise becomes clearly noticeable.

---

###  At t = 200
- `alpha_bar ≈ 0.4387`
- Less than **50% of the original image remains**.
- Image is heavily corrupted by noise.

---

### At t = 499 (final step)
- `alpha_bar ≈ 0.0063`
- Less than **1% of original signal remains**.
- The image is now **pure noise**.

---